nomad/job_endpoint_hooks.go

```diff
@@ -307,45 +308,95 @@ func (v *jobValidate) Validate(job *structs.Job) (warnings []error, err error) {
 
 	for _, tg := range job.TaskGroups {
 		for _, s := range tg.Services {
-			serviceWarn, serviceErr := v.validateServiceIdentity(s)
-			if serviceErr != nil {
-				multierror.Append(validationErrors, serviceErr)
-			}
-			if len(serviceWarn) > 0 {
-				warnings = append(warnings, serviceWarn...)
-			}
+			serviceWarns, serviceErrs := v.validateServiceIdentity(s, fmt.Sprintf("task group %s", tg.Name))
+			multierror.Append(validationErrors, serviceErrs)
+			warnings = append(warnings, serviceWarns...)
 		}
 
 		for _, t := range tg.Tasks {
 			for _, s := range t.Services {
-				serviceWarn, serviceErr := v.validateServiceIdentity(s)
-				if serviceErr != nil {
-					multierror.Append(validationErrors, serviceErr)
-				}
-				if len(serviceWarn) > 0 {
-					warnings = append(warnings, serviceWarn...)
-				}
+				serviceWarns, serviceErrs := v.validateServiceIdentity(s, fmt.Sprintf("task %s", t.Name))
+				multierror.Append(validationErrors, serviceErrs)
+				warnings = append(warnings, serviceWarns...)
 			}
+
+			vaultWarns, vaultErrs := v.validateVaultIdentity(t)
+			multierror.Append(validationErrors, vaultErrs)
+			warnings = append(warnings, vaultWarns...)
 		}
 	}
 
 	return warnings, validationErrors.ErrorOrNil()
 }
 
-func (v *jobValidate) validateServiceIdentity(s *structs.Service) (warnings []error, err error) {
+func (v *jobValidate) validateServiceIdentity(s *structs.Service, parent string) (warnings []error, err error) {
+	var mErr *multierror.Error
+
+	// Prefix errors and warnings.
+	defer func() {
+		prefix := fmt.Sprintf("Service %s in %s", s.Name, parent)
+		for i, w := range warnings {
+			warnings[i] = fmt.Errorf("%s %s", prefix, w)
+		}
+		err = multierror.Prefix(mErr, prefix)
+	}()
```

Comment:
Isn't `warnings` always empty in this method? If so, why not use `fmt.Errorf` for the errors and avoid this extra loop, and then call `multierror.Prefix` before the return?

Reply:
Yeah, it's not used. I debated about not even having them (YAGNI), but I wasn't sure if we would need additional validation as the work progresses.

But back to YAGNI, I will remove this 👍